In [1]:
import pandas as pd
import re


# 01 verify every quote link is in csv for download


This verification was done in the file `03_verify_index_links.ipynb` and the results show that _all the links in the index are in the csv_ file for download.


# 02 Verify every link has been downloaded and every file exists


The file with the link are `acm_site.csv`, `missionary.csv` and `stdhndbk.csv`


In [2]:
# get the files "acm_site.csv" and "missionary.csv" and merge them into a single dataframe
df_exist = pd.read_csv("../data/data_16_09_24/index/all_links.csv")

# sort by filename
df_exist = df_exist.sort_values(by="filename")

df_exist.head(10)


URL  \
499  https://www.byupathway.edu/policies/handbook/6...   
294  https://missionaries.prod.byu-pathway.psdops.c...   
181  https://missionaries.prod.byu-pathway.psdops.c...   
271  https://missionaries.prod.byu-pathway.psdops.c...   
311  https://missionaries.prod.byu-pathway.psdops.c...   
73   https://missionaries.prod.byu-pathway.psdops.c...   
421  https://studentsupportkb.byupathway.org/knowle...   
315  https://missionaries.prod.byu-pathway.psdops.c...   
535                    https://www.englishconnect.org/   
199  https://missionaries.prod.byu-pathway.psdops.c...   

                                               Section  \
499                             ['6. Student Records']   
294  ['Increasing Awareness', 'Shepherding / Minist...   
181                            ['PathwayConnect (PC)']   
271                     ['Missionary Software & Uses']   
311                     ['Missionary Software & Uses']   
73                          ['Calendars', 'Calendars']   
421                         ['Certificates & Degrees']   
315             ['Missionary Software & Uses', 'PATH']   
535                          ['English Connect 1 & 2']   
199                                          ['Block']   

                                            Subsection  \
499                                        ['Missing']   
294  ['Increasing Awareness of PathwayConnect', 'Mi...   
181             ['PathwayConnect General Information']   
271                                       ['Power BI']   
311  ['PATH for Area Coordination Missionaries (ACM...   
73   ['Calendar', 'EnglishConnect 3 Academic Calend...   
421                                      ['Admission']   
315  ['PATH for Area Coordination Missionaries (ACM...   
535                            ['EnglishConnect Home']   
199                                        ['Missing']   

                                                 Title  filename  
499                       ['6.4 Pathway Certificates']  105adb7a  
294  ['Shepherd with Charity', 'Shepherd with Chari...  110f0051  
181                    ['How Students Repeat Courses']  12180468  
271                                        ['PowerBI']  123bdf7c  
311  ['Transferring a Student to a Different Program']  12a1fef2  
73              ['EC3 Auto-Drop', 'Auto-Drop for EC3']  130b6f2b  
421  ['Is there an exception request link for onlin...  1376e9a9  
315  ['Understanding the Groups Page EC3', 'Underst...  13c81944  
535                            ['General Information']  141279b1  
199  ['International Area Transitioning to Block a ...  143661b8

In [3]:
# read the names of the files in the directory "data/data_16_09_24/crawl/html" and "data/data_16_09_24/crawl/pdf"
import os

html_files = os.listdir("../data/data_16_09_24/crawl/html")
pdf_files = os.listdir("../data/data_16_09_24/crawl/pdf")

# ignore txt files that exist in md and txt with the same name in the "../data/data_16_09_24/crawl/pdf"

all_files = html_files + pdf_files

print("HTML files:", len(html_files))
print("PDF files:", len(pdf_files))
print("\n")
print("TOtal files:", len(html_files) + len(pdf_files), "plus 1 files as image.")


HTML files: 246
PDF files: 294


TOtal files: 540 plus 1 files as image.


In [4]:
# function to validate if the file exist
def file_exist(name):
    for file in all_files:
        if name == file.split(".")[0]:
            return True
    return False


In [5]:
# Create a new column called "File Exists" that indicates if the file exists in the directory
df_exist["File Exists"] = df_exist["filename"].apply(file_exist)

# print only the rows where the file does not exist
df_exist[~df_exist["File Exists"]]


URL  \
277  https://missionaries.prod.byu-pathway.psdops.c...   
8            https://degreeapplication.byupathway.edu/   

                           Section               Subsection  \
277  ['Shepherding / Ministering']              ['Missing']   
8       ['Certificates & Degrees']  ['Application Process']   

                                                Title  filename  File Exists  
277     ['Resources for New and Continuing Students']  269ed2e7        False  
8    ['How do students access the BYUI application?']  d26452c9        False

277.- This file is an image, is weird that is not loaded

8.- the link is broken


In [6]:
# print len of existing files
print("Existing files:", len(df_exist[df_exist["File Exists"]]))


Existing files: 540


In [7]:
# Create an Array with the names of the files that do not exist
files_not_exist = df_exist[~df_exist["File Exists"]]["filename"].values


print(len(files_not_exist), "files do not exist")


2 files do not exist


132 files in the list doesn't exist in the download folder.


# 03 Verify if every file in quotes has been downloaded


In [8]:
filename = "../data/temporary/Index_quotes.csv"

df = pd.read_csv(
    filename,
    header=1,
)
df.index = range(1, len(df) + 1)

df.dropna(subset=["Link to Ideal Answer"], inplace=True)

# drop columns "Quotes", "Quote 2" and "Quote 3"
df.drop(columns=["Quotes", "Quote 2", "Quote 3"], inplace=True)

df_no_links = df[df["Link to Ideal Answer"].str.contains("http") == False]

# make a drop of the rows that have no link to ideal answer
df.drop(df_no_links.index, inplace=True)
df.head()


Questions  \
2      How do I know if a student has a scholarship?   
3  How do I know if a student is registered for a...   
5  How do I know if student is member of the church?   
6  What information should I track for each student?   
7  What to do If a student has already taken this...   

                                        Ideal Answer  \
2  Missionaries can see if a student has a schola...   
3  There is not a way for Missionaries to verify ...   
5  Missionaries can see the Church membership sta...   
6  The most important things Missionaries should ...   
7  For a student that is in a course that they ha...   

                                Link to Ideal Answer  \
2  https://missionaries.prod.byu-pathway.psdops.c...   
3  https://missionaries.prod.byu-pathway.psdops.c...   
5  https://missionaries.prod.byu-pathway.psdops.c...   
6  https://missionaries.prod.byu-pathway.psdops.c...   
7  https://pathway-missionary.powerappsportals.co...   

                                              Link 2  Link 3  
2                                                NaN     NaN  
3                                                NaN     NaN  
5                                                NaN     NaN  
6                                                NaN     NaN  
7  https://missionaries.prod.byu-pathway.psdops.c...     NaN

In [9]:
# Función para limpiar y extraer los links
def extract_links(text):
    # si el valor es NaN, retornar una lista vacía
    if pd.isnull(text):
        return []

    # Dividir el texto por los saltos de línea y filtrar solo los enlaces
    links = [
        line.strip() for line in text.split("\n") if line.strip().startswith("https")
    ]
    return links


# Recopila los link de "Link to Ideal Answer", "Link 2" y "Link 3" en un solo campo
def merge_links(row):
    links = row["Link to Ideal Answer"] + row["Link 2"] + row["Link 3"]
    return links


# because some questions has multiple links, we will split them into multiple rows with its respective question
def split_links(row):
    links = row["Links"]
    questions = row["Questions"]
    rows = []
    for link in links:
        rows.append([questions, link])
    return rows


In [10]:
# some of the rows has multiple links, we will convert them to a list
df["Link to Ideal Answer"] = df["Link to Ideal Answer"].apply(extract_links)
df["Link 2"] = df["Link 2"].apply(extract_links)
df["Link 3"] = df["Link 3"].apply(extract_links)

df["Links"] = df.apply(merge_links, axis=1)

list_of_links = df[["Questions", "Links"]]

# Apply the function to the dataframe and convert the result to a dataframe
df_links = list_of_links.apply(split_links, axis=1)
df_links = pd.DataFrame(df_links.sum(), columns=["Questions", "Links"])
df_links.head()


Questions  \
0      How do I know if a student has a scholarship?   
1  How do I know if a student is registered for a...   
2  How do I know if student is member of the church?   
3  What information should I track for each student?   
4  What to do If a student has already taken this...   

                                               Links  
0  https://missionaries.prod.byu-pathway.psdops.c...  
1  https://missionaries.prod.byu-pathway.psdops.c...  
2  https://missionaries.prod.byu-pathway.psdops.c...  
3  https://missionaries.prod.byu-pathway.psdops.c...  
4  https://pathway-missionary.powerappsportals.co...

Now with the links, verify if every of this resources has been downloaded.


In [11]:
# compare the Links from df_links to URL of df_exist, also File Exist must be True


def validate_if_exist(row):
    link = row["Links"]
    # verify if it exist in df_exist dataframe and its colum "File Exists" is True
    exist = df_exist[(df_exist["URL"].isin([link])) & (df_exist["File Exists"] == True)]
    return len(exist) > 0


df_links["File Exists"] = df_links.apply(validate_if_exist, axis=1)


In [12]:
# show all false values
df_some = df_links[~df_links["File Exists"]]

df_some.head()

# show all true values
# df_links[df_links["File Exists"]]


Questions  \
107  How do students access the BYUI application?   

                                         Links  File Exists  
107  https://degreeapplication.byupathway.edu/        False

The origin folder is `data/data_16_09_24/index`

About the file above:

107. Link is broken.


# 04. Validate if Files were transformed to Markdown


In [13]:
import os


In [14]:
# origin

html_files = os.listdir("../data/data_16_09_24/crawl/html")
pdf_files = os.listdir("../data/data_16_09_24/crawl/pdf")

# remove the extension of the files
html_files = [file.split(".")[0] for file in html_files]
pdf_files = [file.split(".")[0] for file in pdf_files]


In [15]:
html_files_out = os.listdir("../data/data_16_09_24/out/from_html")
pdf_files_out = os.listdir("../data/data_16_09_24/out/from_pdf")

# drop txt files
html_files_out = [file for file in html_files_out if file.endswith(".md")]
pdf_files_out = [file for file in pdf_files_out if file.endswith(".md")]

# remove the extension of the files
html_files_out = [file.split(".")[0] for file in html_files_out]
pdf_files_out = [file.split(".")[0] for file in pdf_files_out]


In [16]:
# verify if the file exists in the directory "out_sep_12/from_html" or "out_sep_12/from_pdf", create a list of the files that do not exist
# rememeber to avoid the extension of the file
dont_exist_html = []
dont_exist_pdf = []

for file in html_files:
    if file.split(".")[0] not in html_files_out:
        dont_exist_html.append(file)

for file in pdf_files:
    if file.split(".")[0] not in pdf_files_out:
        dont_exist_pdf.append(file)


print("HTML files that do not exist:", len(dont_exist_html))
print("PDF files that do not exist:", len(dont_exist_pdf))


HTML files that do not exist: 0
PDF files that do not exist: 1


In [17]:
print("len of dont_exist_pdf:", len(dont_exist_pdf))
print("\n")
print(dont_exist_pdf)


len of dont_exist_pdf: 1


['95e946eb']


WARNING! This file, is a PDF file, and for a strange reason, when is converted to txt, it has many weird characters that prevents to run the parser.
